specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+### Why are messages length-prefixed with a protobuf varint in the SSZ encoding?
+
+In stream-oriented protocols, we need to delimit messages from one another, so that the reader knows where one message ends and the next one starts. Length-prefixing is an effective solution. Alternatively, one could set a delimiter char/string, but this can readily cause ambiguity if the message itself may contain the delimiter. It also introduces another set of edge cases to model for, thus causing unnecessary complexity, especially if messages are to be compressed (and thus mutated beyond our control).
+
+That said, in our case, streams are single-use. libp2p streams are full-duplex, and each party is responsible for closing their write side (like in TCP). We therefore use stream closure to mark the end of a request.
+
+Nevertheless, messages are still length-prefixed to prevent DOS attacks where malicious actors send large amounts of data disguised as a request. A length prefix allows clients to set a maximum limit, and once that limit is read, the client can cease reading and disconnect the stream. This allows a client to determine the exact length of the packet being sent, and it capacitates it to reset the stream early if the other party expresses they intend to send too much data.
+
+[Protobuf varint](https://developers.google.com/protocol-buffers/docs/encoding#varints) is an efficient technique to encode variable-length ints. Instead of reserving a fixed-size field of as many bytes as necessary to convey the maximum possible value, this field is elastic in exchange for 1-bit overhead per byte.
+
+### Why do we version protocol strings with ordinals instead of semver?
+
+Using semver for network protocols is confusing. It is never clear what a change in a field, even if backwards compatible on deserialisation, actually implies. Network protocol agreement should be explicit. Imagine two peers:
+
+-  Peer A supporting v1.1.1 of protocol X.
+-  Peer B supporting v1.1.2 of protocol X.
+
+These two peers should never speak to each other because the results can be unpredictable. This is an oversimplification: imagine the same problem with a set of 10 possible versions. We now have 10^2 (100) possible outcomes that peers need to model for. The resulting complexity is unwieldy.
+
+For this reason, we rely on negotiation of explicit, verbatim protocols. In the above case, peer B would provide backwards compatibility by supporting and advertising both v1.1.1 and v1.1.2 of the protocol.
+
+Therefore, semver would be relegated to convey expectations at the human level, and it wouldn't do a good job there either, because it's unclear if "backwards-compatibility" and "breaking change" apply only to wire schema level, to behaviour, etc.
+
+For this reason, we remove semver out of the picture and replace it with ordinals that require explicit agreement and do not mandate a specific policy for changes.
```

Comment:
It's unclear to me what the versioning policy is - is it strict equality (`==`) or are protocols by default backwards compatible (`<=`)? IMO, this should be explicit in the spec.

Reply:
It's strict equality. 
Support for previous version require the local node to explicitly support a version. You can order which is attempted to be negotiated first, i.e higher version before lower ones.

Reply:
Indeed. I’ll clarify/reword this in the spec.

Reply:
514ff5814f233836eba8b45d9bb5d94b97d7a60e